### PR TITLE
Chapel updates post release

### PR DESCRIPTION
**Pull request by arezaii**
Fixes and updates after Chapel's 2.1.0 release. 

# Change Highlights:
(thank you @bonachea!)
* Fix `+rocm` variant, to ensure correct dependencies on `ROCm` packages and use of AMD LLVM
* Add a `+pshm` variant for `comm=gasnet` to enable fast shared-memory comms between co-locales
* Add logic to ensure we get the native CXI `libfabric` network provider on Cray EX
* Expand dependency type for package modules to encompass runtime dependencies
* Factor logic for setting `(LD_)LIBRARY_PATH` and `PKG_CONFIG_PATH` of runtime dependencies
* Workaround issue #44746 that causes a transitive dependency on `lua` to break SLURM
* Disable nonfunctional `checkChplDoc` test
* Annotate some variants as conditional, to improve `spack info` output and reduce confusion

### CODE REVIEW

**Comment by arezaii** [comment]
I don't understand the CI check output from `spack -d audit packages`. It seems to complain that there's a missing `^None`, but adding that and retesting fails the same test and says it's currently using `None ^None` and should instead be using `^None ^None`... 

I also tried making it `^libfabric libfabric=spack` and it gave me the same complaint about using `None ^libfabric` and suggested it should be `^None ^libfabric`, which also did not work and seems like nonsense. 

```console
SystemExit: 1
PKG-DIRECTIVES: 1 issue found
1. chapel: wrong 'when=' condition in 'requires' directive
    using 'None libfabric=spack', should be '^None libfabric=spack'
```

Any ideas what is actually wrong with this block?
```python
    with when(is_CrayEX()):
        requires(
            "^libfabric fabrics=cxi",
            when="libfabric=spack",
            msg="libfabric requires cxi fabric on HPE-Cray EX machines",
        )
```

**Comment by arezaii** [comment]
> Any ideas what is actually wrong with this block?
```python
    with when(is_CrayEX()):
        requires(
            "^libfabric fabrics=cxi",
            when="libfabric=spack",
            msg="libfabric requires cxi fabric on HPE-Cray EX machines",
        )
```
the change that satisfied the `audit` command was swapping the two `when` conditions:
```python
    with when("libfabric=spack"):
        requires(
            "^libfabric fabrics=cxi",
            when=is_CrayEX(),
            msg="libfabric requires cxi fabric on HPE-Cray EX machines",
        )
```

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 1]
```diff
@@ -277,11 +287,28 @@ class Chapel(AutotoolsPackage, CudaPackage, ROCmPackage):
...
+        multi=False,
+        when="comm=gasnet comm_substrate=ofi",
+    )
+
+    with when("libfabric=spack"):
+        requires(
+            "^libfabric fabrics=cxi",
+            when=is_CrayEX(),
+            msg="libfabric requires cxi fabric on HPE-Cray EX machines",
+        )
```
Comment (bonachea) [inline var/spack/repos/builtin/packages/chapel/package.py]: I suspect we want `spack.platforms.cray.slingshot_network()` here instead of `is_CrayEX()`, because the condition in question here is whether we detect a Slingshot-11 network (which is a slightly different question from whether this is a Cray EX).

Suggested change, along with `import spack.platforms.cray` in the header:
```suggestion
    requires(
        "^libfabric" + (" fabrics=cxi" if spack.platforms.cray.slingshot_network() else ""),
        when="libfabric=spack",
        msg="libfabric requires cxi fabric provider on HPE-Cray EX machines",
    )
```

**Review by bonachea — CHANGES_REQUESTED** [review]

**Review by bonachea — APPROVED** [review]
LGTM! I'm excited to see these improvements merged.

Passes my testing using various install configurations on both NERSC Perlmutter and two x86_64/InfiniBand clusters (one with AMD/ROCm GPUs).

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 2]
```diff
@@ -410,17 +436,34 @@ class Chapel(AutotoolsPackage, CudaPackage, ROCmPackage):
     conflicts("platform=windows")  # Support for windows is through WSL only
 
     conflicts("+rocm", when="+cuda", msg="Chapel must be built with either CUDA or ROCm, not both")
-    conflicts("+rocm", when="@:2.0.0", msg="ROCm support in spack requires Chapel 2.0.0 or later")
+    conflicts(
+        "+rocm", when="@:1.99.99", msg="ROCm support in spack requires Chapel 2.0.0 or later"
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:440]: (minor) you can express this as `when=@:1`

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 3]
```diff
@@ -361,7 +387,7 @@ class Chapel(AutotoolsPackage, CudaPackage, ROCmPackage):
             ),
             default=True,
         )
-        depends_on(dep, when="+{0}".format(variant_name))
+        depends_on(dep, when="+{0}".format(variant_name), type=("build", "link", "run", "test"))
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:390]: (minor) `test` should be covered by the other deptypes, it is typically useful when used alone. There's no harm in adding it here (and if you found you *had* to also specify it, I'd be curious to know where).

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 4]
```diff
@@ -410,17 +436,34 @@ class Chapel(AutotoolsPackage, CudaPackage, ROCmPackage):
...
+    )
+
+    # comm_substrate=udp gasnet_segment=unset defaults to everything,
+    # which is incompatible with +pshm
+    requires("gasnet_segment=fast", when="+pshm comm_substrate=udp")
+
     conflicts(
         "^python@3.12:",
-        when="@:2.1.0",
+        when="@:2.0.99",
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:466]: (minor) similar to other comment, this can be expressed as `when=@:2.0`.

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 5]
```diff
@@ -523,13 +570,25 @@ def setup_chpl_compilers(self, env):
...
         # the C/C++ compilers must work post-install
-        if self.spec.satisfies("+cuda") or self.spec.satisfies("+rocm"):
+        if self.spec.satisfies("+rocm"):
+            env.set("CHPL_TARGET_COMPILER", "llvm")
+            env.set(
+                "CHPL_LLVM_CONFIG",
+                join_path(self.spec["llvm-amdgpu"].prefix, "bin", "llvm-config"),
+            )
+            real_cc = join_path(self.spec["llvm-amdgpu"].prefix, "bin", "clang")
+            real_cxx = join_path(self.spec["llvm-amdgpu"].prefix, "bin", "clang++")
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:580]: Note: `llvm-amdgpu` and `llvm` are providers of `libllvm`, so you might be able to consolidate this and the next `elif` by referring to `spec['libllvm']` (not entirely though since it looks like the statements after this are just for `+rocm`

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 6]
```diff
@@ -562,12 +618,18 @@ def prepend_cpath_include(self, env, prefix):
         if not is_system_path(prefix):
             env.prepend_path("CPATH", prefix.include)
 
+    def set_lib_path(self, env, prefix):
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:621]: (minor) I'd recommend "update_" vs. "set_"

File: var/spack/repos/builtin/packages/chapel/package.py [hunk 7]
```diff
@@ -578,47 +640,51 @@ def setup_env_vars(self, env):
...
+        if not self.spec.satisfies("llvm=none"):
+            # workaround Spack issue #44746:
+            # Chapel does not directly utilize lua, but many of its
+            # launchers depend on system installs of batch schedulers
+            # (notably Slurm on Cray EX) which depend on a system Lua.
+            # LLVM includes lua as a dependency, but a barebones lua
+            # install lacks many packages provided by a system Lua,
+            # which are often required by system services like Slurm.
+            # Disable the incomplete Spack lua package directory to
+            # allow the system one to function.
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/chapel/package.py:652]: (minor) that issue suggests that adding `depends_on(lua-luaposix)` would solve this problem, although generally I'm unsure if that's the ideal solution:

If I understand, the issue is that these external scheduler executables check for `LUA_PATH` and use it. Ideally we'd have some way to do that whenever the external scheduler is used (vs. for every package that happens to use both the scheduler and depend on lua).

This seems acceptable as a short term fix though.

**Review by scheibelp — APPROVED** [review]
I have a number of minor comments. The most interesting is to do some consolidation of LLVM inspection. Overall I don't think any of them should block merging (but I do recommend changing some of those things in a future PR).

**Comment by arezaii** [comment]
Thank you @scheibelp! As requested we will revisit and address your concerns above in a future PR

File: var/spack/repos/builtin/packages/chapel/package.py (hunk 7 above)
Comment (bonachea) [reply var/spack/repos/builtin/packages/chapel/package.py:652]: I agree this is an ugly hack and I wish there was a better solution.

FWIW the `lua-luaposix` package alone was insufficient to solve the immediate problem (it just changed one missing lua package error into a second one). IIUC SLURM's plugin architecture means a system SLURM install can depend on a site-specific number of additional lua packages the sysadmins may have installed. Tracking all those potential dependencies in a Chapel package seems inappropriate, since Chapel itself does not directly depend on lua at all. Chapel simply wants to invoke the system SLURM commands and not have them drop dead due to missing packages in Spack's underprovisioned lua install.

Perhaps what we actually want is something like `depends_on("scheduler")`, where `scheduler` is a virtual dependency fulfilled by any number of (external system-provided) scheduler packages that also correctly encode all their transitive dependencies with variants (e.g. a list of specific lua packages). But unless those can be automatically detected with very smart external detection, that likely places the burden on the user running Spack to correctly express all those dependencies.
